include/dr/mhp/algorithms/inclusive_scan.hpp

```diff
@@ -0,0 +1,305 @@
+// SPDX-FileCopyrightText: Intel Corporation
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#pragma once
+
+#ifdef SYCL_LANGUAGE_VERSION
+#include <oneapi/dpl/async>
+#endif
+
+namespace dr::mhp {
+namespace __detail {
+
+template <class FunT>
+inline Event for_each_local_async(auto &&first, auto &&last, FunT &&func) {
+  Event foreach_event;
+#ifdef SYCL_LANGUAGE_VERSION
+  if (use_sycl()) {
+    foreach_event = oneapi::dpl::experimental::for_each_async(
+        dpl_policy(), first, last, std::forward<FunT>(func));
+  } else
+#endif
+  {
+    std::for_each(std::execution::par_unseq, first, last,
+                  std::forward<FunT>(func));
+  }
+  return foreach_event;
+}
+
+template <class FunT>
+inline Event inclusive_scan_local_async(auto &&first, auto &&last,
+                                        auto &&dst_first, FunT &&func) {
+  Event scan_event;
+#ifdef SYCL_LANGUAGE_VERSION
+  if (mhp::use_sycl()) {
+    scan_event = oneapi::dpl::experimental::inclusive_scan_async(
+        dpl_policy(), first, last, dst_first, func);
+  } else {
+    std::inclusive_scan(std::execution::par_unseq, first, last, dst_first,
+                        func);
+  }
+#else
+  std::inclusive_scan(std::execution::par_unseq, first, last, dst_first, func);
+#endif
+  return scan_event;
+}
+
+template <class T>
+inline Event copy_value_after_event(const T *src, T *dst, Event prerequisite) {
+  Event e;
+#ifdef SYCL_LANGUAGE_VERSION
+  if (mhp::use_sycl()) {
+    e = mhp::sycl_queue().submit([&](auto &&h) {
+      h.depends_on(prerequisite);
+      h.single_task([=]() {
+        T tmp = *src;
+        *dst = tmp;
+      });
+    });
+  } else
+#endif
+  {
+    *dst = *src;
+  }
+  return e;
+}
+
+// move to sycl_support.hpp if need to reuse
+inline void wait_for_events_and_clear(std::vector<Event> &events) {
+#ifdef SYCL_LANGUAGE_VERSION
+  sycl::event::wait(events);
+#endif
+  events.clear();
+}
+
+// inclusive scan algorithm:
+// let's denote segments as: rank(dot)segment_number
```

Comment:
This is not the algorithm that I expected: https://courses.cs.washington.edu/courses/cse332/21sp/lectures/19-ParallelPrefix.pdf

I expected up pass to be partial sums and down pass to be scan. But instead there is scan followed by adding a scalar to a vector. Doesn't that require 25% more memory bandwidth?

Reply:
Thanks for the pdf. It will be definitely a nice part of my job to learn this new alg. I basically did what is in shp (with a bit more flexible way).

Could you please write what is the most important case we target our system:
1) number of processes << (much lower) number of elements in one process
or 
2) number of processes ~ number of elements in one process
or
3) number of processes >> number of elements in one process
?

I guess the 1st choice.

Reply:
1. is the common case.

I see that you followed the SHP algorithm. I was wondering if there is some issue with using sycl device memory. @BenBrock: any comments?

Reply:
If there is only 1 segment for the entire distributed vector, then it should do an inclusive scan and not the other work. Otherwise, single rank DR will be 2x slower than standard implementation. Not sure what your implementation does in this case.

For gbench, you need micros that show baseline performance  for a single rank that do not use DR. One for SYCL using DPL, one for CPU using std. Verify that single rank DR matches conventional implementation on 1 rank, and reasonable scalability for multiple ranks with DR.

Reply:
good point, my implementation was slower in this case - I've fixed it. Now I use inclusive_scan with init value from the very beginning when computing first non-empty local sum (segment) and skip for_each for this partial sum (segment).
See fixed most recent version.

Reply:
You still have the issue for the parallel case. A reduce + scan is faster than a scan + vector add. You are memory bound and the vector add uses 2x memory bandwidth compared to reduce.

Reply:
right, I'm fixing it

Reply:
The issue is fixed. Now I use reduce + scan except case of one-segment where just scan is used.
I'd like to test performance, but I've a problem with running mhp-bench with MPI (see slack discussion).

Code looks good to me and is ready to be re-reviewed. Could you please have a look? (unless you want to wait for performance results first)